Allow for configuration of OJ options

conorevans opened the pull request:
Signed-off-by: Conor Evans <coevans@tcd.ie>

**Which issue(s) this PR fixes**: 
It relates to #3311 but does not close it 

**What this PR does / why we need it**: 

> > is there a way to change DEFAULT_OJ_OPTIONS variable ?

> It seems there is no way to do it (pull request is welcome 😄)

from @ashie 

**Docs Changes**:
- TODO: add docs for this

**Release Note**: 

Hi @ashie, I saw the discussion on #3311 and figured I could take this on. I just wanted to loop you in first:

- TODO: add documentation for these env vars
- TODO: add testing to types explicitly (it is somewhat tested via env_utils)
- TODO: should I add support for all of https://github.com/ohler55/oj/blob/develop/pages/Options.md? I didn't want to create a huge PR before consulting, so I just set the current options (+ `max_nesting` as requested on #3311 by @bazzilio)

conorevans commented on lib/fluent/env_utils.rb:
```diff
@@ -0,0 +1,50 @@
+      OJ_OPTIONS_DEFAULTS = {
+        'bigdecimal_load': :float,
+        'mode': :concat,
+        'use_to_json': true
```
worth noting that this existing default is not actually doing anything according to the docs:

> :use_to_json [Boolean]

> Call to_json() methods on dump, default is false. The option is ignored in the :compat and :rails modes.

**The option is ignored in the :compat mode**

conorevans commented on lib/fluent/config/types.rb:
```diff
@@ -104,6 +108,16 @@ def self.string_value(val, opts = {}, name = nil)
+    def self.symbol_value(val, opts = {}, name = nil)
+      return nil if val.nil?
+
+      val[0] == ":" ? val[1..-1].to_sym : val.to_sym
```
I was somewhat surprised that Ruby wouldn't account for this in `to_sym` 🤷 

```zsh
ruby abc.rb
"x"
":bigdecimal"
"x.to_sym"
:":bigdecimal"
"x[1..-1].to_sym"
:bigdecimal
```

conorevans replied:
Done 👍

conorevans commented:
I'm also happy to extend the `max_nesting` functionality beyond `oj` (i.e. to `yajl` which is actually used) but I wasn't sure if these two things should be done in the same PR. Let me know - thanks.

cosmo0920 commented on lib/fluent/config/types.rb:
```diff
@@ -176,6 +196,7 @@ def self.enum_value(val, opts = {}, name = nil)
         when :bool then Config.bool_value(value, opts, name)
         when :time then Config.time_value(value, opts, name)
         when :regexp then Config.regexp_value(value, opts, name)
+        when :symbol then SYMBOL_TYPE.call(value, opts, name)
```
> it could simple be `Config.symbol_value` here

I prefer to use `Config.symbol_value` here. Thanks for your work! :+1:

ashie replied:
+1

conorevans replied:
Done 👍

ashie commented on lib/fluent/env_utils.rb:
```diff
@@ -0,0 +1,50 @@
+
+module Fluent
+  module EnvUtils
+    def self.convert_env_var_to_type(value, type)
```
Are this method and `EnvUtils` modules needed?
It seems using `Fluent::Config.reformatted_value` directly is enough.
And moving OjOptions to just under `Fluent` is more simple.

ashie commented on lib/fluent/env_utils.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    class OjOptions
+      OJ_OPTIONS = {
```
I think they are enough for now.
I want to keep YAGNI.

ashie commented on lib/fluent/config/types.rb:
```diff
@@ -130,7 +144,13 @@ def self.enum_value(val, opts = {}, name = nil)
       else
-        val.to_i
+        i = val.to_i
+        # string will convert to 0, return nil if not actually "0"
```
You are right, but it breaks backward compatibility (several tests are broken by this change).
Please see the just former block. `opts[:strict]` is prepared for this issue.
The default value of `:strict` may be changed to `true` at Fluentd 2.0 but can't do it for now.

ashie replied:
e.g. https://github.com/fluent/fluentd/pull/3315/checks?check_run_id=2264359093

```
2021-04-04T13:34:41.4151745Z Failure: test: not assumed case[hoge](TestConfigTypes::Config.size_value)
2021-04-04T13:34:41.4154661Z /home/runner/work/fluentd/fluentd/test/config/test_types.rb:26:in `block (2 levels) in <class:TestConfigTypes>'
2021-04-04T13:34:41.4155628Z      23:          "hoge" => [0, "hoge"],
2021-04-04T13:34:41.4156173Z      24:          "empty" => [0, ""])
2021-04-04T13:34:41.4157004Z      25:     test 'not assumed case' do |(expected, val)|
2021-04-04T13:34:41.4157811Z   => 26:       assert_equal(expected, Config.size_value(val))
2021-04-04T13:34:41.4158454Z      27:     end
2021-04-04T13:34:41.4158884Z      28: 
2021-04-04T13:34:41.4159507Z      29:     test 'nil' do
2021-04-04T13:34:41.4160038Z <0> expected but was
2021-04-04T13:34:41.4160567Z <nil>
```

conorevans replied:
Right, thanks for the pointer. Addressed 👍

ashie commented on lib/fluent/env_utils.rb:
```diff
@@ -0,0 +1,50 @@
+          env_value = ENV["FLUENT_OJ_OPTION_#{key.upcase}"]
+          next if env_value.nil?
+
+          cast_value = Fluent::EnvUtils.convert_env_var_to_type(env_value, OJ_OPTIONS[key])
```
Please consider adding `{strict: true}` as third argument, as I described above.

conorevans replied:
Done 👍

ashie reviewed (changes requested):
Sorry for my late response.
It's basically O.K. for me but some changes are required (see inline comments).

> I'm also happy to extend the max_nesting functionality beyond oj (i.e. to yajl which is actually used) but I wasn't sure if these two things should be done in the same PR. Let me know - thanks.

I think it should be another PR.

kenhys commented:
https://github.com/fluent/fluentd/pull/3315/commits/a6623bc692f5c109aca242cc1a21f8f7c2a1c934
https://github.com/fluent/fluentd/pull/3315/commits/020693aa2dec3acea117d6d8bd389db801130305
https://github.com/fluent/fluentd/pull/3315/commits/0d7c98008732a8e5631b433d36d0328f9bdc9012
It is better to squash these commits in one commit.

kenhys commented on lib/fluent/env_utils.rb:
```diff
@@ -0,0 +1,50 @@
+        OJ_OPTIONS_DEFAULTS.each { |key, value| @options[key] = value }
+      end
+
+      def get_options
```
`get_options` may be a redundant name.
It seems that `options` is enough.

conorevans replied:
Would that not clash with `@options` above?

kenhys replied:
It seems that it does not conflict with it.

conorevans commented:
Thanks for feedback. Aim to get back to this soon.

conorevans commented:
Will aim to get back to this at some point in 7-14 days, cannot before then unfortunately.

kenhys commented:
@conorevans we are looking forward to make this PR forward!

conorevans commented:
Hi @kenhys sorry this slipped off my mind, thanks for the reminder!

I've addressed the feedback left on the MR. I think squashing the commits is fine (isn't that an option on merge? or shall I clean that up manually?).

kenhys commented:
> I think squashing the commits is fine (isn't that an option on merge? or shall I clean that up manually?).

Yes, PR should be done with a fineness of commit effectively (supporting Config.symbol_value and test. Oj options and test case so on)

ashie commented:
Sorry for my response.
I confirmed that the things I pointed out are fixed. Thanks for your work!

But CI was failed due to errors like following:

https://pipelines.actions.githubusercontent.com/ZYtSfNSOzFZbJsnI7WpW4orkGZmyuPUw7mkWiXZKPSFdBSfK7l/_apis/pipelines/1/runs/2212/signedlogcontent/3?urlExpires=2021-07-05T07%3A56%3A28.4329674Z&urlSigningMethod=HMACV1&urlSignature=29l%2Bas9fKsaFJ1%2BPikZ4CahzSImmK5Dg%2BsLwOPrZ25Q%3D
```
2021-07-05T07:19:30.6421804Z Failure: test: all secret parameters in worker section is sealed(TestFluentdCommand::config dump):
2021-07-05T07:19:30.6423828Z   unexpected error in launching fluentd: #<Errno::ENOENT: No such file or directory - bundle>.
2021-07-05T07:19:30.6424612Z   <"">
2021-07-05T07:19:30.6425163Z   was expected to include:
2021-07-05T07:19:30.6425743Z   <["secret xxxxxx"]>
2021-07-05T07:19:30.6426291Z   and not include:
2021-07-05T07:19:30.6426895Z   <["secret secret0", "secret secret1"]>
2021-07-05T07:19:30.6428005Z /home/runner/work/fluentd/fluentd/test/command/test_fluentd.rb:162:in `assert_log_matches'
2021-07-05T07:19:30.6429405Z /home/runner/work/fluentd/fluentd/test/command/test_fluentd.rb:1072:in `block (2 levels) in <class:TestFluentdCommand>'
```

It's caused on loading lib/fluent/config.rb and its dependencies.
But it seems that it's caused by unnecessary dependencies.
So I add a commit to remove such dependency.

ashie commented:
> So I add a commit to remove such dependency.

Hmm, although it solves the issue on my environment, not yet solved on CI environment :thinking: : https://github.com/fluent/fluentd/pull/3315/checks?check_run_id=2988122775

ashie commented:
> > So I add a commit to remove such dependency.
> 
> Hmm, although it solves the issue on my environment, not yet solved on CI environment thinking : https://github.com/fluent/fluentd/pull/3315/checks?check_run_id=2988122775

I've removed this commit since it's doesn't take effect on it.

ashie commented on test/test_oj_options.rb:
```diff
@@ -0,0 +1,31 @@
+
+  sub_test_case "OjOptions" do
+    test "when no env vars set, returns default options" do
+      ENV.clear
```
Ah, now I've got the reason why so many tests are broken.
Because `ENV` is cleared here, later tests which depend on `ENV` are broken.
I'll fix it.

conorevans replied:
Ah I see, nice catch. Thanks

ashie commented:
I'll fix broken tests and add trivial fixes.
After finishing them, I'll merge this.

ashie commented:
@kenhys Could you review my additional commit especially 3217fe4da5641c47f42ca9f5e29fd2341fdff1e5 ?
Because `Oj.default_options` shouldn't be set by each plugins, I modified to set it only once at starting up of fluentd.

kenhys commented:
Ok, I'll do it.

kenhys reviewed (approved):
It seems that load_env works as expected. :+1:

ashie commented:
Thanks for your contribution!

conorevans commented:
No problem @ashie - thanks for helping close it out!
